File: src/test/java/de/zalando/zally/rules/SnakeCaseInPropNameRuleTest.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import io.swagger.models.ModelImpl;
import io.swagger.models.Swagger;
import io.swagger.models.properties.StringProperty;
import org.junit.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class SnakeCaseInPropNameRuleTest {

    SnakeCaseInPropNameRule rule = new SnakeCaseInPropNameRule();

    private Swagger testSwagger = new Swagger();
    private ModelImpl testDefinition1 = new ModelImpl();
    private ModelImpl testDefinition2 = new ModelImpl();
    private StringProperty testPorperty1 = new StringProperty();
    private StringProperty testPorperty2 = new StringProperty();

    @Test
    public void validateEmptyPath() {
        assertThat(rule.validate(testSwagger)).isEmpty();
    }

    @Test
    public void validateNormalProperty() {
        testDefinition1.addProperty("test_property", testPorperty1);
        testSwagger.addDefinition("ExampleDefintion", testDefinition1);
        assertThat(rule.validate(testSwagger)).isEmpty();
    }

    @Test
    public void validateMultipleNormalProperties() {
        testDefinition1.addProperty("test_property", testPorperty1);
        testDefinition1.addProperty("test_property_two", testPorperty2);
        testSwagger.addDefinition("ExampleDefintion", testDefinition1);
        assertThat(rule.validate(testSwagger)).isEmpty();
    }

    @Test
    public void validateMultipleNormalPropertiesInMultipleDefinitions() {
        testDefinition1.addProperty("test_property", testPorperty1);
        testDefinition2.addProperty("test_property_two", testPorperty2);
        testSwagger.addDefinition("ExampleDefintion", testDefinition1);
        testSwagger.addDefinition("ExampleDefintionTwo", testDefinition2);
        assertThat(rule.validate(testSwagger)).isEmpty();
    }

    @Test
    public void validateFalseProperty() {
        testDefinition1.addProperty("TEST_PROPERTY", testPorperty1);
        testSwagger.addDefinition("ExampleDefintion", testDefinition1);
        List<Violation> result = rule.validate(testSwagger);
        assertThat(result.size()).isEqualTo(1);
        assertThat(result.get(0).getDescription()).contains("\n Violating property: TEST_PROPERTY in definition: ExampleDefintion");
    }

    @Test
    public void validateMultipleFalsePropertiesInMultipleDefinitions() {
        testDefinition1.addProperty("TEST_PROPERTY", testPorperty1);
        testDefinition2.addProperty("test_property_TWO", testPorperty2);
        testSwagger.addDefinition("ExampleDefintion", testDefinition1);
        testSwagger.addDefinition("ExampleDefintionTwo", testDefinition2);
        List<Violation> result = rule.validate(testSwagger);
        assertThat(result.size()).isEqualTo(2);
        assertThat(result.get(0).getDescription()).contains("\n Violating property: TEST_PROPERTY in definition: ExampleDefintion");
        assertThat(result.get(1).getDescription()).contains("\n Violating property: test_property_TWO in definition: ExampleDefintionTwo");
    }
}

File: server/src/test/java/de/zalando/zally/rules/SnakeCaseInPropNameRuleTest.kt
package de.zalando.zally.rules

import de.zalando.zally.swaggerWithDefinitions
import io.swagger.models.ModelImpl
import io.swagger.models.Swagger
import io.swagger.models.properties.StringProperty
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class SnakeCaseInPropNameRuleTest {

    @Test
    fun emptySwagger() {
        assertThat(SnakeCaseInPropNameRule().validate(Swagger())).isNull()
    }

    @Test
    fun validateNormalProperty() {
        val swagger = swaggerWithDefinitions("ExampleDefinition" to listOf("test_property"))
        assertThat(SnakeCaseInPropNameRule().validate(swagger)).isNull()
    }

    @Test
    fun validateMultipleNormalProperties() {
        val swagger = swaggerWithDefinitions("ExampleDefinition" to listOf("test_property", "test_property_two"))
        assertThat(SnakeCaseInPropNameRule().validate(swagger)).isNull()
    }

    @Test
    fun validateMultipleNormalPropertiesInMultipleDefinitions() {
        val swagger = swaggerWithDefinitions(
                "ExampleDefinition" to listOf("test_property"),
                "ExampleDefinitionTwo" to listOf("test_property_two")
        )
        assertThat(SnakeCaseInPropNameRule().validate(swagger)).isNull()
    }

    @Test
    fun validateFalseProperty() {
        val swagger = swaggerWithDefinitions("ExampleDefinition" to listOf("TEST_PROPERTY"))
        val result = SnakeCaseInPropNameRule().validate(swagger)!!
        assertThat(result.description).contains("TEST_PROPERTY")
        assertThat(result.paths).hasSameElementsAs(listOf("#/definitions/ExampleDefinition"))
    }

    @Test
    fun validateMultipleFalsePropertiesInMultipleDefinitions() {
        val swagger = swaggerWithDefinitions(
                "ExampleDefinition" to listOf("TEST_PROPERTY"),
                "ExampleDefinitionTwo" to listOf("test_property_TWO")
        )
        val result = SnakeCaseInPropNameRule().validate(swagger)!!
        assertThat(result.description).contains("TEST_PROPERTY", "test_property_TWO")
        assertThat(result.paths).hasSameElementsAs(listOf(
                "#/definitions/ExampleDefinition",
                "#/definitions/ExampleDefinitionTwo")
        )
    }

    @Test
    fun notFireOnWhitelistedProperty() {
        val swagger = swaggerWithDefinitions("ExampleDefinition" to listOf("_links"))
        assertThat(SnakeCaseInPropNameRule().validate(swagger)).isNull()
    }
}
